### DISCUSSION TITLE
Difference between an interface and contract

### DISCUSSION THREAD

**Discussion by Balogunsamuel** [Cyfrin/foundry-full-course-cu#200] [General]
During the creation of a contract we use ERC20 contract or IERC20, what is the requirement/reason for using an interface and not the main contract every time.

**Comment by thurendous** [comment]
It is totally fine by using a main contract and not the interface of it. But we usually use the interface only, because: 
1. Less code: Main contract can be more than we need. 
2. More understandable: Interface is plain and simple to understand, so devs feel good. 
3. Less gas: If you have less code, it means less gas cost during deployment. 

Also the definition and feature of Interface and main contract are different. 
E.g. An ERC20 interface is respected by the say ERC20 standard usually. But the implementation of the function can be different. -> Interface exists for this reason also.

**Comment by akhilmanga** [comment]
@Balogunsamuel We can use main contract also, but we use only interface. Because the interface code will be small, clear. The computer will understand clearly that these are the main functions , doing certain task in the contract. main functions, errors are included in the interface contract. Solidity allows you to interact with other contracts without having their code by using their interface.

Use interface in the code, to interact with other contracts.

Interface declaration and implementation code: 

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.17;

contract Counter {
    uint public count;

    function increment() external {
        count += 1;
    }
}

interface ICounter {
    function count() external view returns (uint);

    function increment() external;
}

contract MyContract {
    function incrementCounter(address _counter) external {
        ICounter(_counter).increment();
    }

    function getCount(address _counter) external view returns (uint) {
        return ICounter(_counter).count();
    }
}

// Uniswap example
interface UniswapV2Factory {
    function getPair(
        address tokenA,
        address tokenB
    ) external view returns (address pair);
}

interface UniswapV2Pair {
    function getReserves()
        external
        view
        returns (uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast);
}

contract UniswapExample {
    address private factory = 0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f;
    address private dai = 0x6B175474E89094C44Da98b954EedeAC495271d0F;
    address private weth = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;

    function getTokenReserves() external view returns (uint, uint) {
        address pair = UniswapV2Factory(factory).getPair(dai, weth);
        (uint reserve0, uint reserve1, ) = UniswapV2Pair(pair).getReserves();
        return (reserve0, reserve1);
    }
}

**Comment by alymurtazamemon** [comment]
@Balogunsamuel 

Smart contracts are the implementation of the program we want to write, it contains all the variables, events, errors, implementations of functions, and many other things. On the other hand, interfaces just tell us about the architecture/structure of the smart contracts but do not contain any implementations.

Let's say `ERC20.sol` has a function implementation of `balanceOf` which tells the balance of any given account.
```solidity
function balanceOf(address account) public view virtual returns (uint256) {
    return _balances[account];
}
```
Now `IERC20.sol` will only contain the structure of the contract;
```solidity
function balanceOf(address account) public view virtual returns (uint256);
```
which tells that the main `ERC20.sol` has a function `balanceOf` which takes `address` in the parameter and returns `uint256` result. Now interface does not care about what this function will do, it only cares about that the caller of this function should pass the appropriate parameters and get the result.

**Now the question is why do we need these?**

So the answer is interacting with other external contracts can be simpler and easy. Users do not have to worry about their implementation in order to use them, they can just add the interface of it and use it in their smart contracts. They can also only add functions or events which are required in their smart contracts, i.e if users only need `approve` and `transferFrom` functions of `ERC20.sol` contract for use then they do not need to add other functions in interface, they can just add these two functions.

Also, it does not mean you cannot use the main contract, you can do it as well and both ways will work.

**Reply by PatrickAlphaC** [reply]
banger

**Comment by Balogunsamuel** [comment]
Nice. thanks
